client/src/main/java/io/spine/client/CommandRequest.java

```diff
@@ -132,28 +173,85 @@ public CommandRequest onStreamingError(ErrorHandler handler) {
      */
     @CanIgnoreReturnValue
     public ImmutableSet<Subscription> post() {
-        ImmutableSet<Subscription> newSubscriptions = doPost();
-        client().subscriptions()
-                .addAll(newSubscriptions);
-        return newSubscriptions;
-    }
-
-    private ImmutableSet<Subscription> doPost() {
-        MultiEventConsumers consumers = builder.build();
-        Client client = client();
-        Command command = client.requestOf(user())
-                                .command()
-                                .create(message);
-        ImmutableSet<Subscription> result =
-                EventsAfterCommand.subscribe(client, command, consumers, streamingErrorHandler);
-        //TODO:2020-04-17:alexander.yevsyukov: Check the returned Ack and throw an exception
-        // in case of problems.
-        client.post(command);
-        return result;
+        PostOperation op = new PostOperation();
+        return op.perform();
     }
 
     @VisibleForTesting
     CommandMessage message() {
         return message;
     }
+
+    /**
+     * Method object for posting a command.
+     */
+    private final class PostOperation {
+
+        private final Command command;
+        private final MultiEventConsumers consumers;
+        private @Nullable ImmutableSet<Subscription> subscriptions;
+
+        private PostOperation() {
+            this.command =
+                    client().requestOf(user())
+                            .command()
+                            .create(message);
+            this.consumers = eventConsumers.build();
+        }
+
+        private ImmutableSet<Subscription> perform() {
+            subscribeToEvents();
+            Ack ack = client().post(command);
+            Status status = ack.getStatus();
+            switch (status.getStatusCase()) {
+                case OK:
+                    return Optional.ofNullable(subscriptions)
+                                   .orElse(ImmutableSet.of());
+                case ERROR:
+                    cancelVoidSubscriptions();
+                    reportErrorWhenPosting(status);
+                    return ImmutableSet.of();
+                case REJECTION:
```

Comment:
Do we ever return a response in this status?

Reply:
I don't think so. It does not make sense in the current async. implementation, and I could not find such cases. Still, we still have this option, so I've covered this as a branch.

We can possibly deprecate it in the `Status` type, and address it in the code gradually.

Reply:
After some further investigation, I've found out that in some cases a `Status` can hold a `rejection`. It cannot happen when the `Status` is a part of acknowledgement for a command. So, the branch is valid, and we cannot deprecate the field.